py3status/modules/uptime.py

```diff
@@ -63,63 +62,29 @@ class Py3status:
 
     def post_config_hook(self):
-        self._decades = self.py3.format_contains(self.format, 'decades')
-        self._years = self.py3.format_contains(self.format, 'years')
-        self._weeks = self.py3.format_contains(self.format, 'weeks')
-        self._days = self.py3.format_contains(self.format, 'days')
-        self._hours = self.py3.format_contains(self.format, 'hours')
-        self._minutes = self.py3.format_contains(self.format, 'minutes')
-        self._seconds = self.py3.format_contains(self.format, 'seconds')
+        self.dict = {}
+        t = ['decades', 'years', 'weeks', 'days', 'hours', 'minutes', 'seconds']
+        s = [315360000, 31536000, 604800, 86400, 3600, 60, 1]
+        for unit, second in zip(t, s):
+            self.dict[unit] = {
+                'start': self.py3.format_contains(self.format, unit),
+                'seconds': second
+            }
 
     def uptime(self):
-        # Units will be computed from bare seconds since timedelta only
-        # provides .days and .seconds anyway. Getting rid of the seconds
-        # part. Keeping the floating point part would make divmod return
-        # floats, and thus would require days/hours/minutes/seconds to be
-        # casted to int before formatting, which would be dirty to handle
-        # since we can't cast None to int.
         with open('/proc/uptime', 'r') as f:
-            up = int(float(f.readline().split()[0]))
-            offset = time() - up
-
-        cache_timeout = decades = years = weeks = days = hours = minutes = seconds = 0
-
-        # Decades
-        if self._decades:
-            decades, up = divmod(up, 315360000)  # 10 years -> decade
-            cache_timeout = 315360000
-        # Years
-        if self._years:
-            years, up = divmod(up, 31536000)  # 365 days -> year
-            cache_timeout = 31536000
-        # Weeks
-        if self._weeks:
-            weeks, up = divmod(up, 604800)  # 7 days -> week
-            cache_timeout = 604800
-        # Days
-        if self._days:
-            days, up = divmod(up, 86400)  # 24 hours -> day
-            cache_timeout = 86400
-        # Hours
-        if self._hours:
-            hours, up = divmod(up, 3600)  # 60 minutes -> hour
-            cache_timeout = 3600
-        # Minutes
-        if self._minutes:
-            minutes, up = divmod(up, 60)  # 60 seconds -> minute
-            cache_timeout = 60
-        # Seconds
-        if self._seconds:
-            seconds = up  # 1000000000 nanoseconds -> second
-            cache_timeout = 1
-
-        uptime = self.py3.safe_format(self.format, dict(
-            decades=decades, years=years, weeks=weeks, days=days,
-            hours=hours, minutes=minutes, seconds=seconds))
+            uptime = {'up': int(float(f.readline().split()[0]))}
+            offset = time() - uptime['up']
+
+        for unit in self.dict:
+            if self.dict[unit]['start']:
```

Comment:
rather than checking here why don't we just not add it to the dict.  And it would also clean up the code more as we could just do `{..., 'minute': 60, 'seconds': 1}`

I think we should call it `self.time_periods` or something clearer like that.

Reply:
Sounds like it would break this feature...
```
Note: If you don't use a placeholder, its value will be carried over
    to the next placeholder. For example, an uptime of 1 hour 30 minutes
    will give you 90 if {minutes} or 1:30 if {hours}:{minutes}.
```

Reply:
It doesn't change the logic.

It just moves where we do the work to `post_config_hook()`

```
self.time_periods = {}
for unit, second in zip(t, s):
    if self.py3.format_contains(self.format, unit):
        self.time_periods[unit] = seconds
```

Reply:
Done. Review now.